lib/AddressHandler.php

```diff
@@ -0,0 +1,189 @@
+
+namespace OCA\VO_Federation;
+
+use OCA\VO_Federation\Backend\GroupBackend;
+use OCA\VO_Federation\Service\ProviderService;
+use OCP\Federation\ICloudIdManager;
+use OCP\HintException;
+use OCP\IGroupManager;
+use OCP\IL10N;
+use OCP\IURLGenerator;
+
+/**
+ * Class AddressHandler - parse, modify and construct federated sharing addresses
+ *
+ * @package OCA\FederatedFileSharing
+ */
+class AddressHandler {
+	/** @var IL10N */
+	private $l;
+
+	/** @var IURLGenerator */
+	private $urlGenerator;
+
+	/** @var ICloudIdManager */
+	private $cloudIdManager;
+
+	/** @var GroupBackend */
+	private $voGroupBackend;
+
+	/** @var ProviderService */
+	private $providerService;
+
+	/** @var IGroupManager */
+	private $groupManager;
+
+	/**
+	 * AddressHandler constructor.
+	 *
+	 * @param IURLGenerator $urlGenerator
+	 * @param IL10N $il10n
+	 * @param ICloudIdManager $cloudIdManager
+	 */
+	public function __construct(
+		IURLGenerator $urlGenerator,
+		IL10N $il10n,
+		ICloudIdManager $cloudIdManager,
+		GroupBackend $voGroupBackend,
+		ProviderService $providerService,
+		IGroupManager $groupManager
+	) {
+		$this->l = $il10n;
+		$this->urlGenerator = $urlGenerator;
+		$this->cloudIdManager = $cloudIdManager;
+		$this->voGroupBackend = $voGroupBackend;
+		$this->providerService = $providerService;
+		$this->groupManager = $groupManager;
+	}
+
+	/**
+	 * split user and remote from federated cloud id
+	 *
+	 * @param string $address federated share address
+	 * @return array [user, remoteURL]
+	 * @throws HintException
+	 */
+	public function splitUserRemote($address) {
+		try {
+			$cloudId = $this->cloudIdManager->resolveCloudId($address);
+			return [$cloudId->getUser(), $cloudId->getRemote()];
+		} catch (\InvalidArgumentException $e) {
+			$hint = $this->l->t('Invalid Federated Cloud ID');
+			throw new HintException('Invalid Federated Cloud ID', $hint, 0, $e);
+		}
+	}
+
+	/**
+	 * generate remote URL part of federated ID
+	 *
+	 * @return string url of the current server
+	 */
+	public function generateRemoteURL() {
+		return $this->urlGenerator->getAbsoluteURL('/');
+	}
+
+	/**
+	 * check if two federated cloud IDs refer to the same user
+	 *
+	 * @param string $user1
+	 * @param string $server1
+	 * @param string $user2
+	 * @param string $server2
+	 * @return bool true if both users and servers are the same
+	 */
+	public function compareAddresses($user1, $server1, $user2, $server2) {
+		$normalizedServer1 = strtolower($this->removeProtocolFromUrl($server1));
+		$normalizedServer2 = strtolower($this->removeProtocolFromUrl($server2));
+
+		if (rtrim($normalizedServer1, '/') === rtrim($normalizedServer2, '/')) {
+			// FIXME this should be a method in the user management instead
+			\OCP\Util::emitHook(
+				'\OCA\Files_Sharing\API\Server2Server',
+				'preLoginNameUsedAsUserName',
+				['uid' => &$user1]
+			);
+			\OCP\Util::emitHook(
+				'\OCA\Files_Sharing\API\Server2Server',
+				'preLoginNameUsedAsUserName',
+				['uid' => &$user2]
+			);
```

Comment:
It feels quite wrong that a method name compareX would emit prelogin hooks, what is the rationnale here?